crypto/src/merkle.rs

```diff
@@ -35,54 +34,64 @@ impl<T: IntoSchema> IntoSchema for MerkleTree<T> {
     }
 }
 
-/// Represents subtree rooted by the current node
+/// Represents a subtree rooted by the current node.
 #[derive(Debug)]
 pub struct Subtree<T> {
-    /// Left subtree
+    /// Left child node.
     left: Box<Node<T>>,
-    /// Right subtree
+    /// Right child node.
     right: Box<Node<T>>,
-    /// Hash of the node
+    /// Hash of this node.
     hash: HashOf<Node<T>>,
 }
 
-/// Represents leaf node
+/// Represents a leaf node.
 #[derive(Debug)]
 pub struct Leaf<T> {
-    /// Hash of the node
+    /// Hash of this node.
     hash: HashOf<T>,
 }
 
-/// Binary Tree's node with possible variants: Subtree, Leaf (with data or links to data) and Empty.
+/// Binary tree node: [`Subtree`], [`Leaf`] (with data or links to data), or `Empty`.
 #[derive(Debug)]
 #[allow(clippy::module_name_repetitions)]
 pub enum Node<T> {
-    /// Node is root of a subtree
+    /// [`Subtree`] node.
     Subtree(Subtree<T>),
-    /// Leaf node
+    /// [`Leaf`] node.
     Leaf(Leaf<T>),
-    /// Empty node
+    /// Empty node.
     Empty,
 }
 
+#[cfg(feature = "std")]
 #[derive(Debug)]
-/// BFS iterator over the Merkle tree
+/// BFS iterator over [`MerkleTree`].
 pub struct BreadthFirstIter<'itm, T> {
-    queue: Vec<&'itm Node<T>>,
+    queue: VecDeque<&'itm Node<T>>,
 }
 
 #[cfg(feature = "std")]
-impl<U> FromIterator<HashOf<U>> for MerkleTree<U> {
-    fn from_iter<T: IntoIterator<Item = HashOf<U>>>(iter: T) -> Self {
-        let mut hashes = iter.into_iter().collect::<Vec<_>>();
-        hashes.sort_unstable();
-        let mut nodes = hashes
+impl<T> FromIterator<HashOf<T>> for MerkleTree<T> {
+    fn from_iter<I: IntoIterator<Item = HashOf<T>>>(iter: I) -> Self {
+        let mut nodes = iter
             .into_iter()
             .map(|hash| Node::Leaf(Leaf { hash }))
             .collect::<VecDeque<_>>();
-        if nodes.len() % 2 != 0 {
+        nodes.make_contiguous().sort_unstable();
```

Comment:
Addressed https://github.com/hyperledger/iroha/pull/2318#discussion_r892530521, but haven't benchmarked

Reply:
am I correct that, in this case. `make_contiguous` won't make any changes to the queue since all the elements are returned sorted from the input iterator

Reply:
which would mean that the `sort_unstable` is unnecessary?

If this is not the case, shouldn't `MerkleTree::IntoIterator` return elements in sorted order?

Reply:
> If this is not the case, shouldn't `MerkleTree::IntoIterator` return elements in sorted order?

looking at it, the answer is no, BFS will not return elements in sorted order for `MerkleTree`. Now I'm wondering, shouldn't the `IntoIterator` return just the leaves? In which case the return type could be just the hash, not the entire `Node` which, I maintain, is an implementation detail

Reply:
- Some sorting is necessary in the construction, otherwise the same set of leaves results in different root hashes.
After consideration I decided not to impl `Ord` for entire `Node` because it's confusing with the BFS iteration order

- I agree `MerkleTree::iter` is only used in `MerkleTree::leaves` at the moment and there would be no drawbacks to combine them, I don't think `FromIterator` and `IntoIterator` should be always symmetrical.
If I made it symmetrical, `BreadthFirstIter` would turn into `LeavesIter`, which would bring little to developers who want to search the tree and do something with it

Reply:
You're right, I said the wrong thing.
Indeed, I resolved the asymmetry between from/into iterator by removing `impl<'itm, T> IntoIterator for &'itm MerkleTree<T>`.

I'll make another suggestion which is probably close to your idea.
I found this would be even better because ordinary trees iterate leaves

Reply:
Done by just [renaming](https://github.com/hyperledger/iroha/pull/2318/commits/9afdc6440381435a07b2cd8ce142497d210ad201).

I considered
```rust
impl<T> IntoIterator for MerkleTree<T> {
    type Item = HashOf<T>;
    type IntoIter = LeafHashIterator<T>;
    //
}
```
, but left as is because an equivalent can be achieved by filtering `BreadthFirstNodeIterator`

Reply:
"ordinary" was overstated. I don't know trees so much